### ISSUE TITLE
TLS certificate when connecting to postgressql db

### ISSUE THREAD

**Issue by EdwardYGLi** [graphile/postgraphile#1437]
I am running into a TLS certification error when trying to connect to a postgressql db hosted on ibm cloud 
```
A serious error occurred when building the initial schema. Exiting because `retryOnInitFail` is not set. Error details:

Error: self signed certificate in certificate chain
    at TLSSocket.onConnectSecure (_tls_wrap.js:1474:34)
    at TLSSocket.emit (events.js:310:20)
    at TLSSocket._finishInit (_tls_wrap.js:917:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:687:12)
```

Please advise.

**Comment by benjie** [comment]
There's a few ways to deal with this:

- Indicate the SSL key to use as part of the query string: add `?ssl=true&sslrootcert=../../data/your-certificate-here.pem` to the connection string, [like we do in Graphile Starter](https://github.com/graphile/starter/blob/c50d8b5ec7ad9743b4ed3682dcd7cd9e99796a04/heroku-setup.template#L101-L102).
- Disable SSL verification of the PostgreSQL connection with `postgraphile(new pg.Pool({ connectionString: '...', ssl: 'no-verify' }), ...)`
- Disable SSL verification within Node (not recommended): set `NODE_TLS_REJECT_UNAUTHORIZED=0` environmental variable
- Probably more options

**Comment by EdwardYGLi** [comment]
Ok thanks for the solutions. 👍

**Comment by internetuser2008** [comment]
I am facing the same issue, via cli or docker container, and not sure if I understand your solution. If I disable SSL in pg_hba it works, the required solution is SSL to enabled. 
my pg_hba
`hostssl all             all             0.0.0.0/0               md5`

```
postgraphile -c "postgres://user1:user1@192.168.113.10:5432/greendb?ssl=true&sslrootcert=/root/rds-combined-ca-bundle.pem"
postgraphile -c "postgres://user1:user1@192.68.113.10:5432/greendb?ssl=true"

docker run --init -p 5000:5000 graphile/postgraphile --connection postgres://user1:user1@192.168.113.10:5432/greendb?ssl=true --schema public
```

**Comment by benjie** [comment]
@internetuser2008 the RDS certificate bundle will only work with RDS; you need to supply the certificate bundle for your own certificate chain, or use one of the other solutions I listed.

**Comment by miktomic** [comment]
@EdwardYGLi Were you able to get it to work. I am struggling with the same problem. I am able to connect just fine with node-pg, but PostGraphile is not able to handle self-signed certs.

**Comment by benjie** [comment]
@miktomic If you can get `node-pg` to work then (assuming you're using library mode) pass your node-pg `Pool` object as the first argument to PostGraphile: `postgraphile(pool, schemas, options)` - we'll just use that pool rather than building our own then.

Make sure you remember to give your pool error handlers!

**Comment by EdwardYGLi** [comment]
> @EdwardYGLi Were you able to get it to work. I am struggling with the same problem. I am able to connect just fine with node-pg, but PostGraphile is not able to handle self-signed certs.

Hey @miktomic it's been a while but i remember skipping ssl verification fixed it for me.

**Comment by adaja** [comment]
I have the same problem and I don't know what I'm doing wrong.
My log:
```
 Error: self signed certificate in certificate chain
    at TLSSocket.onConnectSecure (node:_tls_wrap:1539:34)
    at TLSSocket.emit (node:events:513:28)
    at TLSSocket._finishInit (node:_tls_wrap:953:8)
    at TLSWrap.ssl.onhandshakedone (node:_tls_wrap:734:12)
```

I'm trying to connect to the PostgreSQL database in AWS RDS.
My docer image is graphile/postgraphile:4.13.0

I make deployment with argument:
`args: [ "--connection", postgres://$(rds_user):$(rds_passwd)@$(rds_endpoint):$(rds_port)/$(rds_dbname)?sslmode=verify-full&sslrootcert=/usr/local/share/ca-certificates/rds-ca-2019-eu-central-1.pem , "--port", "5433", "--schema", "$(graphql_schema)", "--retry-on-init -failed" ]`

I have also check ?ssl=true&sslrootcert=/usr/local/share/ca-certificates/rds-ca-2019-eu-central-1.pem"

`args: [ "--connection", "postgres://$(rds_user):$(rds_passwd)@$(rds_endpoint):$(rds_port)/$(rds_dbname)?ssl=true&sslrootcert=/usr/local/share/ca-certificates/rds-ca-2019-eu-central-1.pem", "--port", "5433", "--schema", "$(graphql_schema)", "--retry-on-init -failed" ]`

and the same issue with pem from https://github.com/graphile/starter/blob/main/data/amazon-rds-ca-cert.pem

However, when I turn off the ssl verification (sslmode=no-verify) , the postgraphile connect to the database.

For the test, I installed postgresql-client in the POD and regardless of whether I provide a certificate or not (with options --set=sslrootcert=/home/node/amazon-rds-ca-cert.pem), I can connect to the database. I see a line like the following in the output: SSL connection (protocol: TLSv1.2, ....)

Am I missing something in the configuration?
Thank you in advance for your help.

**Comment by adaja** [comment]
I have fix this issue:

1) Add env. in deployment: 
```
            - name: NODE_OPTIONS
              value: --use-openssl-ca
```

2) Add eu-central-1-bundle.pem to /usr/local/share/ca-certificates
```
    wget https://truststore.pki.rds.amazonaws.com/eu-central-1/eu-central-1-bundle.pem -P /usr/local/share/ca-certificates   
    update-ca-certificates
```
3) Connection parameter:
`?ssl=true&sslrootcert=/etc/ssl/cert.pem`

**Comment by csbrown** [comment]
I am getting this same error with RDS certificates.  `graphile-migrate watch` succeeds with my $DATABASE_URL to which I've appended `?ssl=true&sslrootcert=../../data/your-certificate-here.pem`.  However, `postgraphile` fails.  It is clearly trying to point to the url in my $DATABASE_URL:

      ‣ Postgres connection: postgres://postgres:[SECRET]@edrr-db-postgres.cjkzjn3eobi3.us-west-2.rds.amazonaws.com/edrr (watching)

but it gives the "self-signed certificate" error.  It's reading the connection string query params, because if I remove them, I get an expected error about no pg_hba.conf entry for my "no encryption" user.  If I delete the `&sslrootcert=..." bit from my connection string, I get the same "self-signed certificate" error.  So, it seems, that `postgraphile` either isn't reading my certificate, or doesn't like my certificate, or something.

I've tried changing the path to my certs to be absolute instead of relative, but that didn't help.
